server/src/test/java/org/elasticsearch/discovery/ClusterDisruptionIT.java

```diff
@@ -436,4 +442,54 @@ public void testIndicesDeleted() throws Exception {
         assertFalse(client().admin().indices().prepareExists(idxName).get().isExists());
     }
 
+    public void testRestartPrimaryNodeWhileIndexing() throws Exception {
+        startCluster(3);
+        String index = "failover_index";
+        assertAcked(client().admin().indices().prepareCreate(index).setSettings(Settings.builder()
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1).put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, between(1, 2))));
+        AtomicBoolean stopped = new AtomicBoolean();
+        Thread[] threads = new Thread[between(1, 4)];
+        AtomicInteger docID = new AtomicInteger();
+        Set<String> ackedDocs = ConcurrentCollections.newConcurrentSet();
+        for (int i = 0; i < threads.length; i++) {
+            threads[i] = new Thread(() -> {
+                while (stopped.get() == false && docID.get() < 5000) {
+                    int docId = frequently() ? docID.getAndIncrement() : between(0, docID.getAndIncrement());
+                    try {
+                        IndexResponse response = client().prepareIndex(index, "_doc", Integer.toString(docId))
+                            .setSource("{\"f\":" + docId + "}", XContentType.JSON).get();
+                        assertThat(response.getResult(), isOneOf(CREATED, UPDATED));
+                        logger.info("--> index id={} seq_no={}", response.getId(), response.getSeqNo());
+                        ackedDocs.add(response.getId());
+                    } catch (ElasticsearchException ignore) {
+                        logger.info("--> fail to index id={}", docId);
+                    }
+                }
+            });
+            threads[i].start();
+        }
+        ensureGreen(index);
+        assertBusy(() -> assertThat(docID.get(), greaterThanOrEqualTo(100)));
+        ClusterState clusterState = internalCluster().clusterService().state();
+        for (ShardRouting shardRouting : clusterState.routingTable().allShards(index)) {
+            if (shardRouting.primary()) {
+                String nodeName = clusterState.nodes().get(shardRouting.currentNodeId()).getName();
+                internalCluster().restartNode(nodeName, new InternalTestCluster.RestartCallback());
```

Comment:
perhaps just restart a random node instead of explicitly the one with the primary?

Reply:
Yep, I pushed https://github.com/elastic/elasticsearch/pull/39584/commits/ab65f4b14ae8af85708e8b4340fd72599546f803